src/main/java/org/openrewrite/java/spring/boot2/LoggingShutdownHooksRecipe.java

```diff
@@ -0,0 +1,76 @@
+package org.openrewrite.java.spring.boot2;
+
+import org.openrewrite.ExecutionContext;
+import org.openrewrite.Recipe;
+import org.openrewrite.TreeVisitor;
+import org.openrewrite.internal.lang.Nullable;
+import org.openrewrite.maven.MavenVisitor;
+import org.openrewrite.maven.tree.Maven;
+import org.openrewrite.maven.tree.Pom;
+import org.openrewrite.xml.XPathMatcher;
+import org.openrewrite.xml.tree.Xml;
+
+import java.util.Collection;
+
+/**
+ * @author Alex Boyko
+ */
+public class LoggingShutdownHooksRecipe extends Recipe {
+
+    private static final XPathMatcher PROJECT_MATCHER = new XPathMatcher("/project");
+
+    private static final String SEARCH_MARKER_DESCRIPTION = "LoggingShutDownHook";
+
+    @Override
+    public String getDisplayName() {
+        return "Logging Shutdown Hooks";
+    }
+
+    @Override
+    public String getDescription() {
+        return "We now register a logging shutdown hook by default for jar based applications to ensure that " +
+                "logging resources are released when the JVM exits. If your application is deployed as a war then " +
+                "the shutdown hook is not registered since the servlet container usually handles logging concerns" +
+                "\n\n" +
+                "Most applications will want the shutdown hook. However, if your application has complex context" +
+                " hierarchies, then you may need to disable it. You can use the logging.register-shutdown-hook" +
+                " property to do that.";
+    }
+
+    @Override
+    protected @Nullable TreeVisitor<?, ExecutionContext> getApplicableTest() {
+        return new MavenVisitor() {
+
+            private String packaging = null;
+
+            @Override
+            public Maven visitMaven(Maven maven, ExecutionContext ctx) {
+                Collection<Pom.Dependency> bootDeps = maven.getModel().findDependencies("org.springframework.boot", "spring-boot");
+                if (bootDeps.stream().filter(d -> d.getVersion().startsWith("2.5.")).findFirst().isPresent()) {
+                    return super.visitMaven(maven, ctx);
+                }
+                return maven;
+            }
+
+            @Override
+            public Xml visitTag(Xml.Tag tag, ExecutionContext context) {
+                if(packaging == null && PROJECT_MATCHER.matches(getCursor())) {
+                    if (tag.getChildValue("packaging").orElse("jar").equalsIgnoreCase("jar")) {
+                        return tag.withMarkers(tag.getMarkers().searchResult());
+                    }
+                }
+                return tag;
+            }
+        };
+    }
+
+    @Override
+    protected TreeVisitor<?, ExecutionContext> getVisitor() {
```

Comment:
Consider using a cursor message provide as an alternative to visitor state variables.  In the following example a cursor message is added to the first Maven parent wich is then removed and inspected when when visiting the Maven tree.
```java
@Override
protected MavenVisitor getVisitor() {
    return new MavenVisitor() {
        @Override
        public Maven visitMaven(Maven maven, ExecutionContext ctx) {
            Maven mvn = super.visitMaven(maven, ctx);
            if (getCursor().pollMessage("NEEDS_SEARCH_MARKER") != null) {
                return mvn.withMarkers(maven.getMarkers().searchResult(SEARCH_MARKER_DESCRIPTION));
            }
            return mvn;
        }
        @Override
        public Xml visitTag(Xml.Tag tag, ExecutionContext context) {
            if(PROJECT_MATCHER.matches(getCursor())) {
                if (tag.getChildValue("packaging").orElse("jar").equalsIgnoreCase("jar")) {
                    getCursor().dropParentUntil(Maven.class::isInstance).putMessage("NEEDS_SEARCH_MARKER", true);
                }
            }
            return tag;
        }
    };
}
```